src/main/java/com/openlattice/datastore/search/controllers/SearchController.java

```diff
@@ -356,40 +352,27 @@ public SearchResult executeFQNPropertyTypeSearch( @RequestBody FQNSearchTerm sea
         List<NeighborEntityDetails> neighbors = Lists.newArrayList();
 
         Set<Principal> principals = Principals.getCurrentPrincipals();
-
-        if ( authorizations.checkIfHasPermissions( new AclKey( entitySetId ), principals,
-                EnumSet.of( Permission.READ ) ) ) {
-            EntitySet es = entitySetManager.getEntitySet( entitySetId );
-
-            checkState( es != null, "Could not find entity set with id: " + entitySetId.toString() );
-
-            final var entitySets = ( es.isLinking() ) ? es.getLinkedEntitySets() : Set.of( entitySetId );
-            final var authorizedEntitySets = entitySets.stream()
-                    .filter( linkedEntitySetId ->
-                            authorizations.checkIfHasPermissions( new AclKey( linkedEntitySetId ),
-                                    Principals.getCurrentPrincipals(),
-                                    EnumSet.of( Permission.READ ) ) )
-                    .collect( Collectors.toSet() );
-            if ( authorizedEntitySets.size() != entitySets.size() ) {
-                logger.warn( "Read authorization failed some of the normal entity sets of linking entity set or it " +
-                        "is empty." );
-            } else {
-                neighbors = searchService
-                        .executeEntityNeighborSearch( ImmutableSet.of( entitySetId ),
-                                new EntityNeighborsFilter( ImmutableSet.of( entityKeyId ) ), principals )
-                        .getOrDefault( entityKeyId, ImmutableList.of() );
-            }
+        if ( getEntitySetsForRead( Set.of( entitySetId ), principals ) != null ) {
+            neighbors = searchGraphService.executeEntityNeighborSearch(
+                    new EntityNeighborsFilterBulk( Map.of( entitySetId, ImmutableSet.of( entityKeyId ) ) ),
+                    principals
+            ).getOrDefault( entityKeyId, ImmutableList.of() );
         }
 
         UUID userId = spm.getCurrentUserId();
 
-        SetMultimap<UUID, UUID> neighborsByEntitySet = HashMultimap.create();
+        Map<UUID, Set<UUID>> neighborsByEntitySet = Maps.newHashMap();
```

Comment:
can use `newHashMapWithExpectedSize( 1.5 * neighbors.size )` (or some other constant >1 depending on how likely the if-statement is)

Reply:
Good call! I have no idea though how likely it is, but probably 1.5 is fine with the 0.75 load factor.